p256/src/arithmetic/field/field32.rs

```diff
@@ -0,0 +1,101 @@
+//! 32-bit secp256r1 field element algorithms.
+
+use super::{MODULUS, MODULUS_HEX};
+use elliptic_curve::bigint::{Limb, U256};
+
+pub(super) const fn add(a: U256, b: U256) -> U256 {
+    let a = a.as_limbs();
+    let b = b.as_limbs();
+
+    // Bit 256 of p is set, so addition can result in nine words.
+    // let (w0, carry) = adc(a[0], b[0], 0);
+    let (w0, carry) = a[0].adc(b[0], Limb::ZERO);
+    let (w1, carry) = a[1].adc(b[1], carry);
+    let (w2, carry) = a[2].adc(b[2], carry);
+    let (w3, carry) = a[3].adc(b[3], carry);
+    let (w4, carry) = a[4].adc(b[4], carry);
+    let (w5, carry) = a[5].adc(b[5], carry);
+    let (w6, carry) = a[6].adc(b[6], carry);
+    let (w7, w8) = a[7].adc(b[7], carry);
+    // Attempt to subtract the modulus, to ensure the result is in the field.
+    let modulus = MODULUS.0.as_limbs();
+
+    let (result, _) = sub_inner(
+        [w0, w1, w2, w3, w4, w5, w6, w7, w8],
+        [
+            modulus[0],
+            modulus[1],
+            modulus[2],
+            modulus[3],
+            modulus[4],
+            modulus[5],
+            modulus[6],
+            modulus[7],
+            Limb::ZERO,
+        ],
+    );
+    U256::new([
+        result[0], result[1], result[2], result[3], result[4], result[5], result[6], result[7],
+    ])
+}
+
+/// Multiplies by a single-limb integer.
+/// Multiplies the magnitude by the same value.
+pub(super) fn mul_single(a: U256, rhs: u32) -> U256 {
+    let mut result = U256::ZERO;
+    for _i in 0..rhs {
+        result = add(a, a)
```

Comment:
This looks like the intent is to add `a` to itself `rhs` times, but I don't think this does that -- it looks like it sets `result` to `a + a` repeatedly, so this will in effect always multiply by 2.

Presumably you wanted this, although I'm not convinced an algorithm that could pontentially run >2^31 adds (since `rhs` is a `u32`) is a good idea.

```suggestion
        result = add(result, a)
```

Reply:
Yes you are right about this. I have changed the current implementation to the suggested change. Although I have another implementation in mind that may solve the second part of your comment. Do let me know what you think about this:
```rust
pub(super) fn mul_single(a: U256, rhs: u32) -> U256 {
    let a = a.as_words();
    let rhs_u64: u64 = rhs as u64;
    U256::from_words([
        a[0] * rhs_u64,
        a[1] * rhs_u64,
        a[2] * rhs_u64,
        a[3] * rhs_u64,
    ])
}
```
for `field64.rs` and
```rust
pub(super) fn mul_single(a: U256, rhs: u32) -> U256 {
    let a = a.as_words();
    U256::from_words([
        a[0] * rhs,
        a[1] * rhs,
        a[2] * rhs,
        a[3] * rhs,
        a[4] * rhs,
        a[5] * rhs,
        a[6] * rhs,
        a[7] * rhs,
    ])
}
```
for `field32.rs`

Reply:
This would only work if there are no carries, and except in special cases there will be carries between the limbs. You could take a similar approach but also track the carries as a separate result, then reduce the carries mod the prime and then do a modular addition of the reduced carries and the non-carries (or I suppose you could skip the "reduce the carries mod the prime" step, but then when you're doing the modular addition you'll have to handle the carry term being a U288 (32-bit) or U320 (64-bit))

Reply:
Since we need to keep track of the carries and mod the prime at the end, I think we can adopt a similar approach to what was implemented for `add`, and replace `adc` with `mac` so that we can track the carries. Then we perform the modulo at the end with `sub_inner` 
```rust
pub(super) fn mul_single(a: U256, rhs: u32) -> U256 {
    let a_limbs = a.as_limbs();
    let rhs_limb = Limb::from_u32(rhs);
    let (w0, carry) = Limb::ZERO.mac(a_limbs[0], rhs_limb, Limb::ZERO);
    let (w1, carry) = Limb::ZERO.mac(a_limbs[0], rhs_limb, carry);
    let (w2, carry) = Limb::ZERO.mac(a_limbs[0], rhs_limb, carry);
    let (w3, w4) = Limb::ZERO.mac(a_limbs[0], rhs_limb, carry);

    // Attempt to subtract the modulus from carry, to ensure the result is in the field
    let modulus = MODULUS.0.as_limbs();

    let (result, _) = sub_inner(
        [w0, w1, w2, w3, w4],
        [modulus[0], modulus[1], modulus[2], modulus[3], Limb::ZERO],
    );
    U256::new([result[0], result[1], result[2], result[3]])
}
```